src/logic/div.rs

```diff
@@ -0,0 +1,1290 @@
+/// - any division that will lead to the quotient being zero or one (no allocation)
+/// - division of any integer by small (1 `Digit`) very small (0.5 `Digit`) integers (no allocation)
+/// - division where the number of leading zeros of both arguments are within one `Digit` (less
+///     allocation than what long division normally requires)
+/// - during long division, the algorithm may encounter a case from above and will use that instead
+/// - division of medium size (<= 512 bits) integers
+/// 
+/// Currently, algorithms faster than 𝒪(n^2) are not implemented, so large integer division may be
+/// very slow compared to other algorithms.
+/// 
+/// Note: Currently there is just one internal division function that is optimized for the
+/// `udivrem` kind of function instead of `uremdiv`. In the future, there is planned a better
+/// implementation for the second one. It was found during designing the first implementation that
+/// this future one should be fundamentally faster than the current one (noticable for small to
+/// medium size `ApInt`s) because of the way `lhs` is subtracted.
+impl ApInt {
+    //Note: the invariant of `ApInt`s where unused bits beyond the bit width must be all zero is
+    //used heavily here, so that no `clear_unused_bits` needs to be used.
+
+    /// This function is intended to be inlined into all of the unsigned quotient and remainder
+    /// functions for optimal assembly.
+    /// `duo` is divided by `div`, and the quotient is assigned to `duo` and remainder assigned
+    /// to `div`
+    /// `false` is returned if division by zero happened. Nothing is modified in the case of
+    /// division by zero.
+    #[inline]
+    pub(crate) fn aarons_algorithm_divrem(duo: &mut [Digit], div: &mut [Digit]) -> bool {
+        //Some parts were put into their own functions and macros because indentation levels were
+        //getting too high, even for me.
+
+        //The algorithm here is just like the algorithm in
+        //https://github.com/AaronKutch/specialized-div-rem,
+        //except that there are more branches and preconditions. There are comments in this function
+        //such as  `//quotient is 0 or 1 check` which correspond to comments in that function.
+        
+        //Divides `duo` by `div` and sets `duo` to the quotient and `div` to the remainder.
+        //Assumptions:
+        // - ini_duo_sd > 0
+        // - div_sd == 0
+        fn large_div_by_small(duo: &mut [Digit], ini_duo_sd: usize, div: &mut [Digit]) {
+            let div_small = div[0];
+            let (mut quo,mut rem) = duo[ini_duo_sd].wrapping_divrem(div_small);
+            duo[ini_duo_sd] = quo;
+            for duo_sd_sub1 in (0..ini_duo_sd).rev() {
+                let duo_double = DoubleDigit::from_lo_hi(duo[duo_sd_sub1],rem);
+                let temp = duo_double.wrapping_divrem(div_small.dd());
+                //the high part is guaranteed to zero out when this is subtracted,
```

Comment:
Please align the comments in the entire crate as I have shown above several times already.